### PR TITLE
Create undertone udapter

### PR DESCRIPTION
**Pull request by gilpat**
## Type of change

- [*] New bidder adapter  
 

## Description of change
new undertone bidder adapter.

- test parameters for validating bids
```
{
   bidder: 'undertone',
          params: {
            placementId: '10433394',
            publisherId: 12345
          }
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- contact email of the adapter’s maintainer -RampProgrammatic@perion.com
- [ ] official adapter submission

### CODE REVIEW

**Comment by gilpat** [comment]
I saw that this error raised while running 'gulp run-tests' 

HeadlessChrome 0.0.0 (Linux 0.0.0) **MarsMedia adapter** implementation "before each" hook FAILED
	Error: Uncaught SyntaxError: Unexpected token < (http://localhost:9876/context.html:1)
HeadlessChrome 0.0.0 (Linux 0.0.0): Executed 859 of 1622 (1 FAILED) (0 secs / 1.241 secs)
HeadlessChrome 0.0.0 (Linux 0.0.0) **MarsMedia adapter** implementation "before each" hook FAILED
	Error: Uncaught SyntaxError: Unexpected token < (http://localhost:9876/context.html:1)
e 0.0.0 (Linux 0.0.0): Executed 859 of 1622 (1 FAILED) (28.686 secs / 1.241 secs)
[15:25:28] 'test-coverage' errored after 3.58 min
[15:25:28] Error: Karma tests failed with exit code 1
    at /home/travis/build/prebid/Prebid.js/gulpfile.js:119:12
    at removeAllListeners (/home/travis/build/prebid/Prebid.js/node_modules/karma/lib/server.js:380:7)
    at Server.<anonymous> (/home/travis/build/prebid/Prebid.js/node_modules/karma/lib/server.js:391:9)
    at Server.g (events.js:291:16)
    at emitNone (events.js:91:20)
    at Server.emit (events.js:185:7)
    at emitCloseNT (net.js:1544:8)
    at _combinedTickCallback (internal/process/next_tick.js:71:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)


I'm not getting this error locally , moreover,  our pull request contains undertone adapter only , nothing that is related to **Marsmedia** .

can you please help?

**Comment by mkendall07** [comment]
@gilpat We have some flakky tests. I've restarted the job to see if it passes

**Comment by gilpat** [comment]
Thanks ,
however the test raises now an error regards to **indexExchange adapter**

**Comment by gilpat** [comment]
Hi , 
is there anything we can do  ? tests seem to fail for no reason I can see .

**Comment by mkendall07** [comment]
@gilpat it's passing now

**Comment by danagutkind** [comment]
Hello Matt,
We would like to go live with prebid adaptor before Dec holidays. 
Is there something that can be done to push forward the approval process of our adapter? 
How many days it usually takes?
We appreciate any help from your side. 
Thank you in advance.

File: modules/undertoneBidAdapter.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    });
+    return {
+      method: 'POST',
+      url: REQ_URL,
+      withCredentials: true,
+      data: JSON.stringify(payload)
+    };
+  },
+  interpretResponse: function(serverResponse, request) {
+    const bids = [];
```
Comment (jsnellbaker) [inline modules/undertoneBidAdapter.js]: #1748 changed the first argument of `interpretResponse` to:

```
{
  body: responseBody,
  headers: {
    get: function(header) { /* returns a header from the HTTP response */ }
  }
}
```

so adding something like

`serverResponse = serverResponse.body;`

just below this line, or however you'd prefer to grab the body, and updating corresponding tests if needed should get this back to working properly

File: modules/undertoneBidAdapter.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    if (serverResponse && Array.isArray(serverResponse) && serverResponse.length > 0) {
+      serverResponse.forEach((bidRes) => {
+        if (bidRes.ad && bidRes.cpm > 0) {
+          const bid = {
+            requestId: bidRes.bidRequestId,
+            bidderCode: BIDDER_CODE,
```
Comment (jsnellbaker) [inline modules/undertoneBidAdapter.js]: `bidderCode` will be set by `bidderFactory` automatically now, this line can be dropped

File: modules/undertoneBidAdapter.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    if (bid && bid.params && bid.params.publisherId && bid.params.placementId) {
+      bid.params.publisherId = parseInt(bid.params.publisherId);
+      return true;
+    }
+  },
+  buildRequests: function(validBidRequests) {
+    const payload = {
+      'x-ut-hb-params': []
+    };
+    const timeout = window.PREBID_TIMEOUT || null;
```
Comment (jsnellbaker) [inline modules/undertoneBidAdapter.js]: Information for prebid can't be accessed through the window object directly.  

As an alternative, you could pass the PREBID_TIMEOUT variable from the page code as another param value like the following:
```
params: {
              placementId: '10433394',
              publisherId: 12345,
              timeout: PREBID_TIMEOUT
              }
```

With this data, you could then access the value within your adapter by reading from the params and assign it to the bid.

**Comment by mkendall07** [comment]
@danagutkind 
if you can get updates on this in the next 30 minutes we can include in the next release.

**Comment by danagutkind** [comment]
@mkendall07  
Thank you very much for your comments. 
We try to make it on time although 30 minutes is a challenge (our guys are at home at this hour).
When is your next release?

File: modules/undertoneBidAdapter.js (hunk 3 above)
Comment (shaymandel) [reply modules/undertoneBidAdapter.js]: Our API enable us to pass the timeout value to the server, so we can response with nobid if we are not able to respond within this time frame. This should improve the user experience and reduce load from our server.
Is there another way to get the timeout you're using? Or maybe there is a constant value you're using? If so, we can just add it here.

**Comment by mkendall07** [comment]
Ok np. Just wanted to give you the heads up. Possibly next week depending on things shake out.

**Comment by danagutkind** [comment]
@mkendall07
Hi Matt,
Furtunatelly, we managed to push all comments from your side. 
We will be grateful if you include out adapter in the closest release. 
Thank you for your help and responsiveness. We appreciate it.

File: modules/undertoneBidAdapter.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
+/**
+ * Adapter to send bids to Undertone
+ */
+
+import * as utils from 'src/utils';
+import { registerBidder } from 'src/adapters/bidderFactory';
+
+const BIDDER_CODE = 'undertone';
+const URL = '//hb.undertone.com/hb'; // //ads.undertone.com/hb
```
Comment (mkendall07) [inline modules/undertoneBidAdapter.js]: What's `//ads.undertone.com/hb` ? Could remove this comment

**Comment by mkendall07** [comment]
@gilpat We are seeing a CORS related error when testing the adapter. Are you sure the test params are valid and work on the hello world test page?

**Comment by jsnellbaker** [comment]
After retesting this morning, we're able to see the test ad with the hello word test page.

LGTM

**Comment by danagutkind** [comment]
@mkendall07
Hello Matt,
Would you please tell when is our adapter expected to go live?
We would like to update our publisher. 
Many thanks again,
Dana

**Comment by danagutkind** [comment]
@mkendall07 Hi Matt
Do you have any estimation when our adapter goes live?
I am asking as we are committed to our publishers.
We would like to reflect timelines as best as we can, so it will be helpful for us to know. 
Can you please advice?

Thank you,
Dana

**Comment by mkendall07** [comment]
What do you mean live? This is merged and available in the 0.34 release I believe

**Comment by danagutkind** [comment]
@mkendall07 

Thank you very much.
I see our adapter is merged and available.

**Pull request merged**